symphony/bdk/core/retry/__init__.py

```diff
@@ -0,0 +1,50 @@
+import logging
+from typing import Callable
+from functools import wraps
+
+from tenacity.retry import retry_if_exception
+from tenacity import stop_after_attempt, wait_exponential, before_sleep_log
+
+from symphony.bdk.core.config.model.bdk_retry_config import BdkRetryConfig
+from symphony.bdk.core.retry.strategy import refresh_session_if_unauthorized
+
+from ._asyncio import AsyncRetrying
+
+
+def retry(*dargs, **dkw):  # noqa
+    """A decorator that provides a mechanism to to retry failed requests
+
+    :param dargs: positional arguments passed to be added or to override the default configuration
+    :param dkw: keyword arguments passed to be added or to override the default configuration
+    """
+    # support both @retry and @retry() as valid syntax
+    if len(dargs) == 1 and callable(dargs[0]):
+        return retry()(dargs[0])
+    else:
+        def retry_decorator(fun: Callable):
+            @wraps(fun)
+            def decorator_f(self, *args, **kwargs):
+                """Fetches a BdkRetryConfiguration object from the instance of the called function
+
+                If no _retry_config attribute is present, use a default AsyncRetrying configuration
+                Passed retry configuration arguments override the default configuration"""
+                default_kwargs = {}
+                retry_config: BdkRetryConfig = getattr(self, '_retry_config', None)
```

Comment:
What happens if `retry_config` is None ? I would prefer to have `retry_config: BdkRetryConfig = self._retry_config` and fail fast with an `AttributeError` in case field does not exist.

Reply:
I will remove the None default value in gettattr and it should raise an AttributeError if the attribute doesn't exist: 
https://docs.python.org/3/library/functions.html#getattr